README
WWW-Google-SiteMap
==================

The Sitemap Protocol allows you to inform search engine crawlers about
URLs on your Web sites that are available for crawling. A Sitemap con-
sists of a list of URLs and may also contain additional information
about those URLs, such as when they were last modified, how frequently
they change, etc.

This module allows you to create and modify sitemaps.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Carp
  XML::Simple
  IO::Zlib
  IO::File

If you want to use the optional feature to notify Google when your sitemaps
are updated, you will also need LWP.

COPYRIGHT AND LICENCE

Copyright (C) 2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension WWW::Google::SiteMap.

1.09
    - Replaced %F with %Y-%m-%d in strftime calls, to correct a Solaris bug
      (since Solaris strftime doesn't know about %F).  Reproted by Fredrik
      Acosta.
    - Added more information to error messages when parsing an existing
      sitemap fails.  Reported by Vinko Vrsalovic Bolte.

1.08
    - Converted installer from ExtUtils::MakeMaker to Module::Install, which
      should make it easier to install for Windows users.
    - Fixed a uri-encoding problem in WWW::Google::SiteMap::URL, reported
      by Jeff Horn.
    - Added some additional test cases and cleaned up some test cases.

1.07
    - Added changes for 1.06 to Changes file, which were accidentally left
      out of the last release.
    - Fixed the WWW-Google-SiteMap-Robot.t test to skip testing if
      WWW::Mechanize is not installed, reported by Jan Kratochvil.

1.06
    - Correction to documentation for WWW::Google::SiteMap::Robot
      ($robot->start should be $robot->run), reported by Martin Kissner.
    - Adjusted WWW::Google::SiteMap::Robot to discard anchors from URLs,
      reported by Martin Kissner.
   
1.05
    - Removed an accidental prerequisite from WWW::Google::SiteMap::URL.  You
      can give it DateTime objects to set the lastmod option, but you don't
      have to, there are other ways to set it as well.

1.04
    - Fixed a documentation error in WWW::Google::SiteMap, and a problem with
      the XML headers generated for sitemap indexes, both reported by
      Michael Smith.
    - Use 'use vars' instead of 'our' for older versions of perl.

1.03
    - Removed some leftover debugging output from WWW::Google::SiteMap::Robot.
    - Fixed date generation to deal with strangeness in the dates that Google
      will accept.  Note that this means some date/time strings that were
      previously accepted may now fail.  See L<WWW::Google::SiteMap::URL> for
      details of the acceptable values to lastmod().
    - Added some more tests.

1.02
    - Fix a problem caused by the switch to XML::Twig, which was leaving off
      some of the XML header information.  Reported by Olaf Anders.
    - Fixed a bug with WWW::Google::SiteMap::Robot, which forgot to load the
      WWW::Google::SiteMap::Ping module before trying to send pings.
 
1.01
    - Accidentally forgot to include the new WWW::Google::SiteMap::Robot
      class, which helps build sitemaps by spidering your web site.

1.00
    - Version 1.00 Released!
    - Fixed Zlib detection problem reported by Lance Cleveland.
    - Check to make sure that the sitemap file was opened correctly, rather
      than just crashing when we try to write to it, also reported by
      Lance Cleveland.
    - Added support for sitemap indexes (see WWW::Google::SiteMap::Index)
    - Added support for notifying Google when your sitemaps and sitemap
      indexes are updated (see WWW::Google::SiteMap::Ping).  Suggested by
      Frank Naude.
    - Fixed a bug in the ISO-8601 time format checking.

0.03
    - Changed from XML::Simple to XML::Twig for XML parsing/generating, this
      means you can now validate your sitemaps with an XML validator.
    - Fixed some documentation errors, spotted by Ing. Branislav Gerzo

0.02
    - Renamed from Google::SiteMap to WWW::Google::SiteMap, shouldn't have
      created a new top-level namespace in the first place.

0.01
	- original version; created by h2xs 1.23 with options
		-X Google::SiteMap



lib/WWW/Google/SiteMap.pm
package WWW::Google::SiteMap;
use vars qw($VERSION); $VERSION = '1.09';

=head1 NAME

WWW::Google::SiteMap - Perl extension for managing Google SiteMaps

=head1 SYNOPSIS

  use WWW::Google::SiteMap;

  my $map = WWW::Google::SiteMap->new(file => 'sitemap.gz');

  # Main page, changes a lot because of the blog
  $map->add(WWW::Google::SiteMap::URL->new(
    loc        => 'http://www.jasonkohles.com/',
    lastmod    => '2005-06-03',
    changefreq => 'daily',
    priority   => 1.0,
  ));

  # Top level directories, don't change as much, and have a lower priority
  $map->add({
    loc        => "http://www.jasonkohles.com/$_/",
    changefreq => 'weekly',
    priority   => 0.9, # lower priority than the home page
  }) for qw(
    software gpg hamradio photos scuba snippets tools
  );

  $map->write;

=head1 DESCRIPTION

The Sitemap Protocol allows you to inform search engine crawlers about URLs
on your Web sites that are available for crawling. A Sitemap consists of a
list of URLs and may also contain additional information about those URLs,
such as when they were last modified, how frequently they change, etc.

This module allows you to create and modify sitemaps.

=cut

use strict;
use warnings;
use WWW::Google::SiteMap::URL qw();
use XML::Twig qw();
unless($IO::Zlib::VERSION) { eval "use IO::Zlib ()"; }
my $ZLIB = $IO::Zlib::VERSION;
use IO::File qw();
require UNIVERSAL;
use Carp qw(carp croak);
use HTML::Entities qw(decode_entities);

=head1 METHODS

=over 4

=item new()

Creates a new WWW::Google::SiteMap object.

  my $map = WWW::Google::SiteMap->new(
    file => 'sitemap.gz',
  );

=cut

sub new {
	my $class = shift;
	my %opts = @_;
	my $self = bless({}, ref($class) || $class);
	while(my($key,$value) = each %opts) { $self->$key($value) }
	if($self->file && -e $self->file) { $self->read }
	return $self;
}

=item read()

Read a sitemap in to this object.  If a filename is specified, it will be
read from that file, otherwise it will be read from the file that was
specified with the file() method.  Reading of compressed files is done
automatically if the filename ends with .gz.

=cut

sub read {
	my $self = shift;
	my $file = shift || $self->file ||
		croak "No filename specified for ".(ref($self)||$self)."::read";

	# don't try to parse missing or empty files
	# no errors for this, because we might be creating it
	return unless -f $file && -s $file;
	
	# don't try to parse very small compressed files
	# (empty .gz files are 20 bytes)
	return if $file =~ /\.gz/ && -s $file < 50;

	my $fh;
	if($file =~ /\.gz$/i) {
		croak "IO::Zlib not available, cannot read compressed sitemaps"
			unless $ZLIB;
		$fh = IO::Zlib->new($file,"rb");
	} else {
		$fh = IO::File->new($file,"r");
	}
	my @urls = ();
	my $urlparser = sub {
		my $self = shift;
		my $elt = shift;

		my $url = WWW::Google::SiteMap::URL->new();
		foreach my $c ($elt->children) {
			my $var = $c->gi;
            if($var eq 'loc') {
                $url->$var(decode_entities($c->text));
            } else {
                $url->$var($c->text);
            }
		}
		$self->purge;
		push(@urls,$url);
	};
	my $twig = XML::Twig->new(
		twig_roots      => {
			'urlset/url'				=> $urlparser,
			'sitemapindex/sitemap'		=> $urlparser,
		},
	);
	$twig->safe_parse(join('',$fh->getlines))
        or die "Could not parse $file ($@)";
	$self->urls(@urls);
}

=item write([$file])

Write the sitemap out to the file.  If a filename is specified, it will be
written to that file, otherwise it will be written to the file that was
specified with the file() method.  Writing of compressed files is done
automatically if the filename ends with .gz.

=cut

sub write {
	my $self = shift;
	my $file = shift || $self->file ||
		croak "No filename specified for ".(ref($self)||$self)."::write";

	my $fh;
	if($file =~ /\.gz$/i) {
		croak "IO::Zlib not available, cannot write compressed sitemaps"
			unless $ZLIB;
		$fh = IO::Zlib->new($file,"wb9");
	} else {
		$fh = IO::File->new($file,"w");
	}
	croak "Could not create '$file'" unless $fh;
	$fh->print($self->xml);
}

=item urls()

Return the L<WWW::Google::SiteMap::URL> objects that make up the sitemap.

=cut

sub urls {
	my $self = shift;
	$self->{urls} = \@_ if @_;
	my @urls = grep { ref($_) && defined $_->loc } @{$self->{urls}};
	return wantarray ? @urls : \@urls;
}

=item add($item,[$item...])

Add the L<WWW::Google::SiteMap::URL> items listed to the sitemap.

If you pass hashrefs instead of L<WWW::Google::SiteMap::URL> objects, it
will turn them into objects for you.  If the first item you pass is a
simple scalar that matches \w, it will assume that the values passed are
a hash for a single object.  If the first item passed matches m{^\w+://}
(i.e. it looks like a URL) then all the arguments will be treated as URLs,
and L<WWW::Google::SiteMap::URL> objects will be constructed for them, but only
the loc field will be populated.

This means you can do any of these:

  # create the WWW::Google::SiteMap::URL object yourself
  my $url = WWW::Google::SiteMap::URL->new(
    loc => 'http://www.jasonkohles.com/',
    priority => 1.0,
  );
  $map->add($url);

  # or
  $map->add(
    { loc => 'http://www.jasonkohles.com/' },
    { loc => 'http://www.jasonkohles.com/software/google-sitemap/' },
    { loc => 'http://www.jasonkohles.com/software/geo-shapefile/' },
  );

  # or
  $map->add(
    loc       => 'http://www.jasonkohles.com/',
    priority  => 1.0,
  );

  # or even something funkier
  $map->add(qw(
    http://www.jasonkohles.com/
    http://www.jasonkohles.com/software/www-google-sitemap/
    http://www.jasonkohles.com/software/geo-shapefile/
    http://www.jasonkohles.com/software/text-fakedata/
  ));
  foreach my $url ($map->urls) { $url->changefreq('daily') }
    
=cut

sub add {
	my $self = shift;
	if(ref($_[0])) {
		if(UNIVERSAL::isa($_[0],"WWW::Google::SiteMap::URL")) {
			push(@{$self->{urls}}, @_);
		} elsif(ref($_[0]) =~ /HASH/) {
			push(@{$self->{urls}},map {
				WWW::Google::SiteMap::URL->new($_)
			} @_);
		}
	} elsif($_[0] =~ /^\w+$/) {
		push(@{$self->{urls}}, WWW::Google::SiteMap::URL->new(@_));
	} elsif($_[0] =~ m{^\w+://}) {
		push(@{$self->{urls}}, map {
			WWW::Google::SiteMap::URL->new(loc => $_)
		} @_);
	} else {
		croak "Can't turn '".(
			ref($_[0]) || $_[0]
		)."' into WWW::Google::SiteMap::URL object";
	}
}

=item xml();

Return the xml representation of the sitemap.

=cut

sub xml {
	my $self = shift;

	my $xml = XML::Twig::Elt->new('urlset', {
		'xmlns'	=> 'http://www.google.com/schemas/sitemap/0.84',
		'xmlns:xsi' => 'http://www.w3.org/2001/XMLSchema-instance',
		'xsi:schemaLocation' => join(' ',
			'http://www.google.com/schemas/sitemap/0.84',
			'http://www.google.com/schemas/sitemap/0.84/sitemap.xsd',
		),
	});
	foreach($self->urls) {
		$_->as_elt->paste(last_child => $xml);
	}
	$xml->set_pretty_print($self->pretty);
	my $header = '<?xml version="1.0" encoding="UTF-8"?>';
	if($self->pretty) { $header .= "\n" }
	return $header.$xml->sprint();
}

=item file()

Get or set the filename associated with this object.  If you call read() or
write() without a filename, this is the default.

=cut

sub file {
	my $self = shift;
	$self->{file} = shift if @_;
	return $self->{file};
}

=item pretty()

Set this to a true value to enable 'pretty-printing' on the XML output.  If
false (the default) the XML will be more compact but not as easily readable
for humans (Google and other computers won't care what you set this to).

If you set this to a 'word' (something that matches /[a-z]/i), then that
value will be passed to XML::Twig directly (see the L<XML::Twig> pretty_print
documentation).  Otherwise if a true value is passed, it means 'nice', and a
false value means 'none'.

Returns the value it was set to, or the current value if called with no
arguments.

=cut

sub pretty {
	my $self = shift;
	my $val = shift || return $self->{pretty} || 'none';

	if($val =~ /[a-z]/i) {
		$self->{pretty} = $val;
	} elsif($val) {
		$self->{pretty} = 'nice';
	} else {
		$self->{pretty} = 'none';
	}
	return $self->{pretty};
}

=back

=head1 MODULE HOME PAGE

The home page of this module is
L<http://www.jasonkohles.com/software/WWW-Google-SiteMap>.  This is where you
can always find the latest version, development versions, and bug reports.  You
will also find a link there to report bugs.

=head1 SEE ALSO

L<WWW::Google::SiteMap::Index>

L<WWW::Google::SiteMap::Ping>

L<WWW::Google::SiteMap::Robot>

L<http://www.jasonkohles.com/software/WWW-Google-SiteMap>

L<https://www.google.com/webmasters/sitemaps/docs/en/protocol.html>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
__END__


lib/WWW/Google/SiteMap/URL.pm
package WWW::Google::SiteMap::URL;
use vars qw($VERSION); $VERSION = '1.09';

=head1 NAME

WWW::Google::SiteMap::URL - URL Helper class for WWW::Google::SiteMap

=head1 SYNOPSIS

  use WWW::Google::SiteMap;

=head1 DESCRIPTION

This is a helper class that supports L<WWW::Google::SiteMap> and
L<WWW::Google::SiteMap::Index>.

=cut

=head1 METHODS

=over 4

=cut

use strict;
use warnings;
use Carp qw(carp croak);
use XML::Twig qw();
use POSIX qw(strftime);
use HTML::Entities qw(encode_entities);

=item new()

=cut

sub new {
	my $class = shift;
	my %opts = ref($_[0]) ? %{$_[0]} : @_;

	my $self = bless({}, $class);
	while(my($key,$value) = each %opts) { $self->$key($value) }
	return $self;
}

=item loc()

Change the URL associated with this object.  For a L<WWW::Google::SiteMap>
this specifies the URL to add to the sitemap, for a
L<WWW::Google::SiteMap::Index>, this is the URL to the sitemap.

=cut

sub loc {
	shift->_doval('loc', sub {
		local $_ = shift;
		return unless defined;
		return 'must be less than 2048 characters long' unless length($_) < 2048;
		return 'must be a fully qualified url' unless m{^https?://};
		return;
	}, @_);
}

=item changefreq()

Set the change frequency of the object.  This field is not used in sitemap
indexes, only in sitemaps.

=cut

sub changefreq {
	shift->_doval('changefreq', sub {
		local $_ = shift;
		my @values = qw(always hourly daily weekly monthly yearly never);
		my $re = join('|',@values);
		return unless defined;
		return 'must be one of '.join(', ',@values) unless /^$re$/;
		return;
	}, @_);
}

=item lastmod()

Set the last modified time.  You have to provide this as one of the following:

=over 4

=item a complete ISO8601 time string

A complete time string will be accepted in exactly this format:

  YYYY-MM-DDTHH:MM:SS+TZ:TZ
  YYYY   - 4-digit year
  MM     - 2-digit month (zero padded)
  DD     - 2-digit year (zero padded)
  T      - literal character 'T'
  HH     - 2-digit hour (24-hour, zero padded)
  SS     - 2-digit second (zero padded)
  +TZ:TZ - Timezone offset (hours and minutes from GMT, 2-digit, zero padded)

=item epoch time

Seconds since the epoch, such as would be returned from time().  If you provide
an epoch time, then an appropriate ISO8601 time will be constructed with
gmtime() (which means the timezone offset will be +00:00).  If anyone knows
of a way to determine the timezone offset of the current host that is
cross-platform and doesn't add dozens of dependencies then I might change this.

=item an ISO8601 date (YYYY-MM-DD)

A simple date in YYYY-MM-DD format.  The time will be set to 00:00:00+00:00.

=item a L<DateTime> object.

If a L<DateTime> object is provided, then an appropriate timestamp will be
constructed from it.

=item a L<HTTP::Response> object.

If given an L<HTTP::Response> object, the last modified time will be
calculated from whatever time information is available in the response
headers.  Currently this means either the Last-Modified header, or tue
current time - the current_age() calculated by the response object.
This is useful for building web crawlers.

=back

Note that in order to conserve memory, any of these items that you provide
will be converted to a complete ISO8601 time string when they are stored.
This means that if you pass an object to lastmod(), you can't get it back
out.  If anyone actually has a need to get the objects back out, then I
might make a configuration option to store the objects internally.

If you have suggestions for other types of date/time objects or formats
that would be usefule, let me know and I'll consider them.

=cut

sub lastmod {
	my $self = shift;

	return $self->{lastmod} unless @_;

	my $value = shift;
	if(ref($value)) {
		if($value->isa('DateTime')) { # DateTime object
			my($date,$tzoff) = $value->strftime("%Y-%m-%dT%T","%z");
			if($tzoff =~ /^([+-])?(\d\d):?(\d\d)/) {
				$tzoff = ($1 || '+').$2.':'.($3||'00');
			} else {
				$tzoff = '+00:00';
			}
			$self->{lastmod} = $date.$tzoff;
		} elsif($value->isa('HTTP::Response')) {
			my $modtime = $value->last_modified()
				|| (time - $value->current_age());
			$self->{lastmod} = strftime("%Y-%m-%dT%T+00:00",gmtime($_));
		}
	} else {
		local $_ = $value;
		if(/^\d+$/) { # epoch time
			$self->{lastmod} = strftime("%Y-%m-%dT%T+00:00",gmtime($_));
		} elsif(/^\d\d\d\d-\d\d-\d\d$/) {
			$self->{lastmod} = $_.'T00:00:00+00:00';
		} elsif(/^\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d\+\d\d:\d\d$/) {
			$self->{lastmod} = $_;
		}
	}

	return $self->{lastmod} if $self->{lastmod};
	$self->_err("'$_' is not a valid value for lastmod");
}

=item priority()

Set the priority.  This field is not used in sitemap indexes, only in sitemaps.

=cut

sub priority {
	shift->_doval('priority', sub {
		local $_ = shift;
		return unless defined;
		return 'must be a number' unless /^[\d\.]+$/;
		return 'must be greater than 0.0' unless $_ >= 0.0;
		return 'must be less than 1.0' unless $_ <= 1.0;
		return;
	}, @_);
}

sub _doval {
	my $self = shift;
	my $var = shift;
	my $valid = shift;
	return $self->{$var} unless @_;
	my $value = shift;
	if(my $res = $valid->($value)) {
		my $msg = "'$value' is not a valid value for $var: $res";
		if($self->{lenient}) { carp $msg } else { croak $msg }
	} else {
		$self->{$var} = $value;
	}
}
sub _err {
	my $self = shift;

	if($self->{lenient}) { carp @_ } else { croak @_ }
}


=item delete()

Delete this object from the sitemap or the sitemap index.

=cut

sub delete {
	my $self = shift;
	for(keys %{$self}) { $self->{$_} = undef }
}

=item lenient()

If lenient contains a true value, then errors will not be fatal.

=cut

sub lenient {
	my $self = shift;
	$self->{lenient} = shift if @_;
	return $self->{lenient};
}

sub as_elt {
	my $self = shift;
	my $type = shift || 'url';
	my @fields = @_;
	unless(@fields) { @fields = qw(loc changefreq lastmod priority) }
	my @elements = ();
	foreach(@fields) {
		my $val = $self->$_() || next;
        if($_ eq 'loc') {
            $val = XML::Twig::Elt->new('#PCDATA' => encode_entities($val));
            $val->set_asis(1);
        } else {
            $val = XML::Twig::Elt->new('#PCDATA' => $val);
        }
        push(@elements,$val->wrap_in($_));
	}
	return XML::Twig::Elt->new($type, {}, @elements);
}

=back

=head1 MODULE HOME PAGE

The home page of this module is
L<http://www.jasonkohles.com/software/WWW-Google-SiteMap>.  This is where you
can always find the latest version, development versions, and bug reports.  You
will also find a link there to report bugs.

=head1 SEE ALSO

L<WWW::Google::SiteMap>

L<WWW::Google::SiteMap::Index>

L<WWW::Google::SiteMap::Ping>

L<http://www.jasonkohles.com/software/WWW-Google-SiteMap/>

L<https://www.google.com/webmasters/sitemaps/docs/en/protocol.html>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
__END__


lib/WWW/Google/SiteMap/Ping.pm
package WWW::Google::SiteMap::Ping;
use vars qw($VERSION); $VERSION = '1.09';

=head1 NAME

WWW::Google::SiteMap::Ping - Perl extension for sending update pings to Google

=head1 SYNOPSIS

  use WWW::Google::SiteMap::Ping;
  
  my $ping = WWW::Google::SiteMap::Ping->new(
    'http://www.jasonkohles.com/sitemap.gz',
  );
  
  $ping->submit;
  print "These pings succeeded:\n";
  foreach($ping->success) {
    print "$_: ".$ping->status($_)."\n";
  }
  print "These pings failed:\n";
  foreach($ping->failure) {
    print "$_: ".$ping->status($_)."\n";
  }

=head1 DESCRIPTION

This module makes it easy to notify Google that your sitemaps, or sitemap
indexes, have been updated.  See L<WWW::Google::SiteMap> and
L<WWW::Google::SiteMap::Index> for tools to help you create sitemaps and
indexes.

=cut

use strict;
use warnings;
use LWP::UserAgent;
use URI::Escape qw(uri_escape);

=head1 METHODS

=over 4

=item new();

Create a new WWW::Google::SiteMap::Ping object.  Can be given a list of
URLs which refer to sitemaps or sitemap indexes, these URLs will simply
be passed to url().

=cut

sub new {
	my $class = shift;
	my $self = bless({}, ref($class) || $class);
	$self->{urls} = {};
	$self->add_urls(@_);
	return $self;
}

=item add_urls(@urls);

Add one or more urls to the list of URLs to submit to Google.

=cut

sub add_urls {
	my $self = shift;
	foreach(@_) {
		$self->{urls}->{$_} ||= 'PENDING';
	}
}

=item urls();

Return the list of urls that will be (or were) submitted to google.

=cut

sub urls { return keys %{shift()->{urls}}; }

=item submit

Submit the urls to Google, returns the number of successful submissions.  This
module uses L<LWP::UserAgent> for the web-based submissions, and will honor
proxy settings in the environment.  See L<LWP::UserAgent> for more information.

=cut

sub submit {
	my $self = shift;

	my $ua = $self->user_agent();
	my $success = 0;
	foreach my $url ($self->urls) {
		my $ping = "http://www.google.com/webmasters/sitemaps/ping?".
			"sitemap=".uri_escape($url);
		my $response = $ua->get($ping);
		if($response->is_success) {
			$self->{urls}->{$url} = 'SUCCESS';
			$success++;
		} else {
			$self->{urls}->{$url} = $response->status_line;
		}
	}

	return $success;
}

=item success();

Return the URLs that were successfully submitted.  Note that success only
means that the request was successfully received by Google, it does not
mean your sitemap was found, loaded or parsed successfully.  If you want
to know whether your sitemap was loaded or parsed successfully, you have
to go to L<http://www.google.com/webmasters/sitemaps> and check the status
there.

=cut

sub success {
	my $self = shift;

	return grep { $self->{urls}->{$_} eq 'SUCCESS' } keys %{$self->{urls}};
}

=item failure();

Return the URLs that were not successfully submitted.

=cut

sub failure {
	my $self = shift;

	return grep { $self->{urls}->{$_} ne 'SUCCESS' } keys %{$self->{urls}};
}

=item user_agent();

If called with no arguments, will return the current L<LWP::UserAgent> object
which will be used to access the web-based submission.  If called with an
arugment, you can set the user agent that will be used in case you need to
give it special arguments.  It must be a L<LWP::UserAgent> object.

If you call submit without having provided a user agent, one will be created
for you that is a basic L<LWP::UserAgent> object, which honors proxy settings
in the environment.

=cut

sub user_agent {
	my $self = shift;

	if(@_) { $self->{_ua} = shift }
	unless($self->{_ua}) {
		$self->{_ua} = LWP::UserAgent->new();
		$self->{_ua}->env_proxy;
		$self->{_ua}->timeout(10);
	}
	return $self->{_ua};
}

=back

=head1 MODULE HOME PAGE

The home page of this module is
L<http://www.jasonkohles.com/software/WWW-Google-SiteMap>.  This is where you
can always find the latest version, development versions, and bug reports.  You
will also find a link there to report bugs.

=head1 SEE ALSO

L<http://www.google.com/webmasters/sitemaps/docs/en/submit.html#ping>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
__END__


lib/WWW/Google/SiteMap/Index.pm
package WWW::Google::SiteMap::Index;
use vars qw($VERSION); $VERSION = '1.09';

=head1 NAME

WWW::Google::SiteMap::Index - Perl extension for managing Google SiteMap Indexes

=head1 SYNOPSIS

  use WWW::Google::SiteMap::Index;
  
  my $index = WWW::Google::SiteMap::Index->new(
    file => 'sitemap-index.gz',
  );
  
  $index->add(WWW::Google::SiteMap::URL->new(
    loc     => 'http://www.jasonkohles.com/sitemap1.gz',
    lastmod => '2005-11-01',
  ));
  
=head1 DESCRIPTION

A sitemap index is used to point Google at your sitemaps if you have more
than one of them.

=cut

use strict;
use warnings;
use base 'WWW::Google::SiteMap';

=head1 METHODS

=over 4

=item new()

Creates a new WWW::Google::SiteMap::Index object.

  my $index = WWW::Google::SiteMap::Index->new(
    file => 'sitemap-index.gz',
  );

=item read()

Read a sitemap index in to this object.  If a filename is specified, it will
be read from that file, otherwise it will be read from the file that was
specified with the file() method.  Reading of compressed files is done
automatically if the filename ends with .gz.

=item write([$file]);

Write the sitemap index out to the file.  If a filename is specified, it will
be written to that file, otherwise it will be written to the file that was
specified with the file() method.  Writing of compressed files is done
automatically if the filename ends with .gz

=item urls()

Return the L<WWW::Google::SiteMap::URL> objects that make up the sitemap index.

=item add($item,[$item...]);

Add the L<WWW::Google::SiteMap::URL> items listed to the sitemap index.

If you pass hashrefs instead of L<WWW::Google::SiteMap::URL> objects, it
will turn them into objects for you.  If the first item you pass is a
simple scalar that matches \w, it will assume that the values passed are
a hash for a single object.  If the first item passed matches m{^\w+://}
(i.e. it looks like a URL) then all the arguments will be treated as URLs,
and L<WWW::Google::SiteMap::URL> objects will be constructed for them, but only
the loc field will be populated.

This means you can do any of these:

  # create the WWW::Google::SiteMap::URL object yourself
  my $url = WWW::Google::SiteMap::URL->new(
    loc => 'http://www.jasonkohles.com/sitemap1.gz',
  );
  $map->add($url);
  
  # or
  $map->add(
    { loc => 'http://www.jasonkohles.com/sitemap1.gz' },
    { loc => 'http://www.jasonkohles.com/sitemap2.gz' },
    { loc => 'http://www.jasonkohles.com/sitemap3.gz' },
  );
  
  # or
  $map->add(
    loc       => 'http://www.jasonkohles.com/sitemap1.gz',
    priority  => 1.0,
  );
  
  # or even something funkier
  $map->add(qw(
    http://www.jasonkohles.com/
    http://www.jasonkohles.com/software/www-google-sitemap/
    http://www.jasonkohles.com/software/geo-shapefile/
    http://www.jasonkohles.com/software/text-fakedata/
  ));
  foreach my $url ($map->urls) { $url->lastmod('2005-11-01') }

=item xml();

Return the xml representation of the sitemap index.

=cut

sub xml {
	my $self = shift;

	my $xml = XML::Twig::Elt->new('sitemapindex', {
		'xmlns' => 'http://www.google.com/schemas/sitemap/0.84',
		'xmlns:xsi' => 'http://www.w3.org/2001/XMLSchema-instance',
		'xsi:schemaLocation' => join(' ',
			'http://www.google.com/schemas/sitemap/0.84',
			'http://www.google.com/schemas/sitemap/0.84/siteindex.xsd',
		),
	});
	foreach($self->urls) {
		$_->as_elt('sitemap',qw(loc lastmod))->paste(last_child => $xml);
	}
	$xml->set_pretty_print($self->pretty);
	my $header = '<?xml version="1.0" encoding="UTF-8"?>';
	if($self->pretty) { $header .= "\n" }
	return $header.$xml->sprint();
}

=item file();

Get or set the filename associated with this object.  If you call read() or
write() without a filename, this is the default.

=item pretty()

Set this to a true value to enable 'pretty-printing' on the XML output.  If
false (the default) the XML will be more compact but not as easily readable
for humans (Google and other computers won't care what you set this to).

If you set this to a 'word' (something that matches /[a-z]/i), then that
value will be passed to XML::Twig directly (see the L<XML::Twig> pretty_print
documentation).  Otherwise if a true value is passed, it means 'nice', and a
false value means 'none'.

Returns the value it was set to, or the current value if called with no
arguments.

=back

=head1 MODULE HOME PAGE

The home page of this module is
L<http://www.jasonkohles.com/software/WWW-Google-SiteMap>.  This is where you
can always find the latest version, development versions, and bug reports.  You
will also find a link there to report bugs.

=head1 SEE ALSO

L<WWW::Google::SiteMap>

L<WWW::Google::SiteMap::Ping>

L<http://www.jasonkohles.com/software/WWW-Google-Sitemap>

L<https://www.google.com/webmasters/sitemaps/docs/en/protocol.html#sitemapFileRequirements>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
__END__
